### ISSUE TITLE
Move safety checks inside combineReducer to first invocation

### ISSUE THREAD

**Issue by gaearon** [rackt/redux#717]
Currently we make a bunch of invariant tests inside `combineReducers()` to make sure it's not being misused in common ways. These tests currently occur at the `combineReducers()` call time.

This makes Redux DevTools unable to catch the errors inside reducers, as they often trigger those invariants to blow up during module definition, and abort hot reloading. Instead, I propose to move the tests to the *first real invocation* of the combined reducer. In most cases that would be `createStore` call because it causes an implicit initial dispatch. This way Redux DevTools can catch and display such errors.

Shall we do that? The code would change to be like

```js
  var reducerSanityVerified;
  var stateShapeVerified;

  // don't do it here

  return function combination(state = defaultState, action) {
    if (!reducerSanityVerified) {
      Object.keys(finalReducers).forEach(key => {
        // do that stuff
```

<img width="729" alt="screen shot 2015-09-13 at 07 09 31" src="https://cloud.githubusercontent.com/assets/810438/9835189/6aa2c77c-59e6-11e5-984f-7fbb19b73692.png">

**Comment by gaearon** [comment]
Another possible alternative is to create errors and `console.error()` them ASAP during `combineReducers()` call, but throw during first dispatch. This way we both have a nice stack and keep DevTools happy.

**Comment by gaearon** [comment]
Would anybody like to work on this?
@ellbee, do you think you might have the time?

**Comment by ellbee** [comment]
Sure, I'll be able to take a look in the next couple of days.

**Comment by ellbee** [comment]
@gaearon I have been working on this, but I am new to the dark arts of HMR and I have a question.

In the following module taken from the counter example, I think delaying the sanity check until the dispatch of the @@INIT action will still blow up the module definition because the init action is getting dispatched either inside `createStore` or `replaceReducer`. This module is going to get re-evaluated too when a reducer changes I think?

```
import React, { Component } from 'react';
import CounterApp from './CounterApp';
import { createStore, applyMiddleware, combineReducers, compose } from 'redux';
import { devTools, persistState } from 'redux-devtools';
import { DevTools, DebugPanel, LogMonitor } from 'redux-devtools/lib/react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import * as reducers from '../reducers';

const finalCreateStore = compose(
  applyMiddleware(thunk),
  devTools(),
  persistState(window.location.href.match(/[?&]debug_session=([^&]+)\b/))
)(createStore);

const reducer = combineReducers(reducers);
const store = finalCreateStore(reducer);

if (module.hot) {
  module.hot.accept('../reducers', () =>
    store.replaceReducer(combineReducers(require('../reducers')))
  );
}

export default class App extends Component {
  render() {
    return (
      <div>
        <Provider store={store}>
          {() => <CounterApp />}
        </Provider>
        <DebugPanel top right bottom>
          <DevTools store={store}
                    monitor={LogMonitor}
                    visibleOnLoad={true} />
        </DebugPanel>
      </div>
    );
  }
}
```

I've got a couple of potential ideas to get around this if it is indeed a problem, just wanted to check that I am not misunderstanding how the HMR works.

**Comment by gaearon** [comment]
It will below up the first time you launch the app but I don't think it's a big deal. It's iteration that I'm concerned about (changing reducers or adding new ones). In case of iteration only the reducer file and replaceReducer() call will be evaluated.

**Comment by ellbee** [comment]
Ah, thanks! That makes total sense. I should read my error messages. The error thrown on iteration is caused by the `Connect` already being hooked up and `mapStateToProps` being triggered by the dispatch in `replaceReducers` with undefined state.

**Comment by gaearon** [comment]
Yeah—that's a different thing, https://github.com/gaearon/redux-devtools/issues/106

**Comment by ellbee** [comment]
So to achieve our goal that devtools issue will need to be addressed too I guess. Having this error thrown from the dispatch in `replaceReducers` is resulting in `[HMR] Cannot apply update. Need to do a full reload!`, so devtools isn't updating and displaying the errors like we want.

**Comment by gaearon** [comment]
Yes, but these are two separate issues. When I tested, fixing them both locally fixed the problem.

**Comment by ellbee** [comment]
Yes, agreed.

**Comment by gaearon** [comment]
Done in #717.

**Comment by gaearon** [comment]
Released in 3.0.1.

**Comment by gaearon** [comment]
There is still a problem with the scenario I described above. **If the reducer throws, then the reducer sanity check itself will throw**. In other words, 

```js
  Object.keys(finalReducers).forEach(key => {
    var reducer = finalReducers[key];
    if (!sanityError && typeof reducer(undefined, { type: ActionTypes.INIT }) === 'undefined') {
```

See that `typeof reducer(...)` call? It will blow up if reducer throws, blowing up `combineReducers` anyway. I'd say that the way sanity check should be changed is the following:

* Let sanity check throw
* Catch the error outside and store as `sanityError`
* Like now, throw it on first dispatch

@ellbee Can you look into this?

**Comment by ellbee** [comment]
Ah yes, I see. Quick fix attempt here #807

**Comment by gaearon** [comment]
Fixed via #807.

**Comment by gaearon** [comment]
Final fix in 3.0.2, confirmed as working.
Thanks again.
